### DISCUSSION TITLE
dict[str, Any] isn't a Mapping[Hashable, Any]?

### DISCUSSION THREAD

**Discussion by ringohoffman** [microsoft/pyright#7306] [Q&A]
```python
from collections.abc import Hashable, Mapping
from typing import Any


def map_arguments(kwargs: Mapping[str, Any], mapper: Mapping[Hashable, Any]) -> dict[str, Any]:
    return {k: mapper[v] if isinstance(v, Hashable) and v in mapper else v for k, v in kwargs.items()}


arg_mapper = {
    "a": "b"
}

map_arguments({"e": "a"}, arg_mapper)
# Argument of type "dict[str, str]" cannot be assigned to parameter "mapper" of type "Mapping[Hashable, Any]" in function "map_arguments"
#   "dict[str, str]" is incompatible with "Mapping[Hashable, Any]"
#     Type parameter "_KT@Mapping" is invariant, but "str" is not the same as "Hashable"
```

What am I missing? `issubclass(str, Hashable) == True`

**Comment by erictraut** [comment]
Pyright is correct in flagging this as an error. `dict[str, Any]` is not a subtype of `Mapping[Hashable, Any]` because the first type parameter of `Mapping` is invariant. The second type parameter of `Mapping` is covariant.

**Comment by Azureblade3808** [comment]
There are methods like `keys` and `items` on `Mapping`, which return the key type, so the key type becomes invariant.

I personally don't think these methods are essential to a map protocol, and I like the key type to be contravariant, so I've written my own protocol. I won't recommend doing so though, unless you have really strong need.
